### ISSUE TITLE
[BUG]: Don't use source to get the mask of IGNORE_INDEX

### ISSUE THREAD

**Issue by gauss-clb** [hpcaitech/ColossalAI#4490]
### 🐛 Describe the bug

Code is here. https://github.com/hpcaitech/ColossalAI/blob/main/applications/Chat/coati/dataset/sft_dataset.py#L40

Don't concatenate source and target, and use the length of source to mask sequences(source + target). Because the last token of source and the first token of target may be merged into one token when concatenating source and target.

For example, 
```python
soucre = 'Human: ... Assistant:_'
target  = '我是...'
```
I use `_` represent whitespace, and when concatenating soucre and target, the token '_' and '我' will be merged into '_我'. 
Therefore, we should tokenize source and target individually, and then merge the sequence of token_ids.

Code may be like,
```python
def _tokenize_fn(user_prompt, completion, tokenizer, max_input_length, max_length, add_eos_token=True):
    tokenizer.add_bos_token = False
    completion_result = tokenizer(
        completion,
        truncation=True,
        max_length=max_length,
        padding=False,
        return_tensors=None,
    )
    tokenizer.add_bos_token = True
    user_prompt_result = tokenizer(
        user_prompt,
        truncation=True,
        max_length=max_input_length,
        padding=False,
        return_tensors=None,
    )
    input_len = len(user_prompt_result["input_ids"])
    result = {}
    result["input_ids"] = user_prompt_result["input_ids"] + completion_result["input_ids"][:max_length-input_len]
    result["attention_mask"] = user_prompt_result["attention_mask"] + completion_result["attention_mask"][:max_length-input_len]
    
    if add_eos_token and len(result["input_ids"]) >= max_length:
        result["input_ids"][max_length - 1] = tokenizer.eos_token_id
        result["attention_mask"][max_length - 1] = 1

    result["labels"] = result["input_ids"].copy()
    result["labels"][:input_len] = [IGNORE_INDEX] * input_len
    return result
```
